webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/queries/LdapQueryProcessor.scala

```diff
@@ -350,20 +350,30 @@ class InternalLDAPQueryProcessor(
     // Returns None if the resulting composition (using AND) leads to an empty set of Nodes, Some filtering otherwise
     def dnMapSets(normalizedQuery: LDAPNodeQuery, dnMapMapSets: Map[DnType, Map[String,Set[DN]]]) : Option[Map[DnType, Set[DN]]] = {
       val mapSet = dnMapMapSets map { case (dnType, dnMapSet) =>
-        val dnSet:Set[DN] = normalizedQuery.composition match {
-          case Or  => dnMapSet.foldLeft(Set[DN]())( _ union _._2 )
+        val dnSet: Option[Set[DN]] = normalizedQuery.composition match {
+          case Or  => dnMapSet.foldLeft(Set[DN]().some){
+            case (Some(a),b) =>  Some(a union b._2 )
+            case (None,b)    =>  Some(b._2)
+          }
           case And =>
-            val s = if(dnMapSet.isEmpty) Set[DN]() else dnMapSet.foldLeft(dnMapSet.head._2)( _ intersect _._2 )
-            // Here if s is empty, it means we are ANDing with an empty Set, so we could simply drop all computation from there
-            if(s.isEmpty && normalizedQuery.transform != ResultTransformation.Invert) {
-              logPure.logEffect.debug(s"[${debugId}] `-> early stop query (non-inverted empty sub-query)")
-              return None
+            dnMapSet.toList match {
```

Comment:
i'm concerned about the convertion from map to list, and then from map to list to map again ( mapSet.toList.sequence.map(_.toMap) ) 
cost of convertion can be prohibitive with thousands of nodes

i don't have a better proposal though :/

Reply:
I think i fixed it here in this PR: https://github.com/fanf/rudder/pull/13 

I never convert the original map, and only fold in an optionnal result map